python/services/evaluationservice/dmod/evaluationservice/runner.py

```diff
@@ -1,30 +1,108 @@
 #!/usr/bin/env python3
+"""
+Script for running a listener that will launch workers to run requested evaluations
+"""
+import collections
+import traceback
 import typing
 import os
 import sys
-import multiprocessing
 import json
 import signal
+import enum
 
 from argparse import ArgumentParser
 
+from concurrent import futures
+from datetime import timedelta
+from functools import partial
+
+from dmod.metrics import CommunicatorGroup
+from dmod.core.context import DMODObjectManager
+from dmod.core.common.collection import TimedOccurrenceWatcher
+
+from dmod.core.context import get_object_manager
+from dmod.core.common.protocols import JobLauncherProtocol
+from dmod.core.common.protocols import JobResultProtocol
+
 import service
 import utilities
 import worker
 
+from service.service_logging import get_logger
+
+MONITOR_DELAY: int = 5
+"""The number of seconds to wait before polling the job queue again"""
+
+
+_ExitCode = collections.namedtuple('ExitCode', ['code', 'explanation'])
+
+
+class _ExitCodes(_ExitCode, enum.Enum):
+    """
+    Exit Codes and their corresponding meanings for this application
+    """
+    SUCCESSFUL = 0, "Naturally Exited"
+    UNEXPECTED = 1, "Exitted Unexpectedly"
+    FORCED = 2, "Forced to exit by signal"
+    TOO_MANY_ERRORS = 3, "Exitted due to too many errors"
+
+    @classmethod
+    def print_codes(cls):
+        """
+        Print out the explanation for each exit code to stdout
+        """
+        for exit_code in cls:
+            print(str(exit_code))
+
+    def exit(self):
+        """
+        Exit the application with the appropriate code
+        """
+        sys.exit(self.code)
+
+    def __str__(self):
+        return f"{self.code}={self.explanation}"
+
+
+def get_concurrency_executor_type(**kwargs) -> typing.Callable[[], JobLauncherProtocol]:
+    """
+    Gets the class type that will be responsible for running evaluation jobs concurrently
+
+    Returns:
+        The type of executor that should be used to run the evaluation jobs
+    """
+    method = os.environ.get("EVALUATION_RUNNER_CONCURRENCY_METHOD", "multiprocessing").lower()
+
+    if method == "threading":
+        return partial(futures.ThreadPoolExecutor, **kwargs)
+
+    return partial(futures.ProcessPoolExecutor, **kwargs)
+
 
 def signal_handler(signum, frame):
+    """
+    Catches and handles signals sent into the application from the os, such as a termination or keyboard interupt
+
+    Args:
+        signum:
+        frame:
+    """
     service.error("Received external signal. Now exiting.")
-    sys.exit(1)
+    _ExitCodes.FORCED.exit()
 
 
-class Arguments(object):
+class Arguments:
```

Comment:
Not a dataclass since the passed arguments are just the arguments sent straight from `sys.argv` through to `argparse.ArgumentParser`. Changing the entire CLI is a tad out of scope, but I'm going to keep it on my radar and mull it about.